### PR TITLE
fix: Return JSON-RPC protocol errors for unknown tools

### PR DESCRIPTION
**Pull request by bzsurbhi**
Return JSON-RPC protocol errors for unknown tools

## Motivation and Context
The MCP server was returning tool execution errors for unknown tools instead of proper JSON-RPC protocol errors. This caused issues where Unknown tools were treated as tool execution failures rather than protocol-level errors

**Before(Tool execution error):**
```
{
  "jsonrpc": "2.0",
  "id": 3,
  "result": {
    "content": [{"type": "text", "text": "Unknown tool: invalid_tool_name"}],
    "isError": true
  }
}
```

**After(Protocol error):**
```
{
  "jsonrpc": "2.0",
  "id": 3,
  "error": {
    "code": -32601,
    "message": "Unknown tool: invalid_tool_name"
  }
}
```

Modified the tool call handler in src/mcp/server/lowlevel/server.py to:
- Check for unknown tools after retrieving from cache
- Raise McpError with METHOD_NOT_FOUND (-32601) error code for unknown tools

## How Has This Been Tested?
- Added unit test
- Verified with a server and client example

## Breaking Changes
This is a breaking change that corrects the error handling behavior:
- Unknown tools now return JSON-RPC protocol errors
- Valid tools that fail during execution still return tool execution errors
- Input validation errors continue to work as before
- Warning logs are still generated for unknown tools during lookup

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

### CODE REVIEW

File: src/mcp/server/lowlevel/server.py [hunk 1]
```diff
@@ -452,6 +452,14 @@ async def handler(req: types.CallToolRequest):
                     arguments = req.params.arguments or {}
                     tool = await self._get_cached_tool_definition(tool_name)
 
+                    # Check if tool exists - return protocol error if not found
+                    if tool is None:
+                        raise McpError(
+                            types.ErrorData(
+                                code=types.METHOD_NOT_FOUND,
```
Comment (ihrpr) [inline src/mcp/server/lowlevel/server.py:459]: Tool name in the request is a parameter (the method is `tools/call`), hence I'd suggest to change to INVALID_PARAMS (-32602) to be consistent with the spec https://modelcontextprotocol.io/specification/2025-06-18/server/tools#error-handling and [Typescript SDK](https://github.com/modelcontextprotocol/typescript-sdk/blob/ihrpr/zod-types/src/server/mcp.ts#L145)

**Review by ihrpr — CHANGES_REQUESTED** [review]
Thank you for the fix! 

Please update the error code to `types.INVALID_PARAMS`. Additionally, consider adding test cases for edge cases like empty or null tool names.

**Comment by maxisbey** [comment]
Closing due to inactivity.

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>
